src/twisted/internet/defer.py

```diff
@@ -886,49 +934,23 @@ def ensureDeferred(coro):
     coroutine. If a Deferred is passed to this function, it will be returned
     directly (mimicing C{asyncio}'s C{ensure_future} function).
 
-    Coroutine functions return a coroutine object, similar to how generators
-    work. This function turns that coroutine into a Deferred, meaning that it
-    can be used in regular Twisted code. For example::
-
-        import treq
-        from twisted.internet.defer import ensureDeferred
-        from twisted.internet.task import react
-
-        async def crawl(pages):
-            results = {}
-            for page in pages:
-                results[page] = await treq.content(await treq.get(page))
-            return results
-
-        def main(reactor):
-            pages = [
-                "http://localhost:8080"
-            ]
-            d = ensureDeferred(crawl(pages))
-            d.addCallback(print)
-            return d
-
-        react(main)
+    See L{Deferred.fromCoroutine} for examples of coroutines.
 
     @param coro: The coroutine object to schedule, or a L{Deferred}.
     @type coro: A Python 3.5+ C{async def} C{coroutine}, a Python 3.4+
         C{yield from} using L{types.GeneratorType}, or a L{Deferred}.
 
     @rtype: L{Deferred}
     """
-    from types import GeneratorType
-
-    if version_info >= (3, 4, 0):
-        from asyncio import iscoroutine
-
-        if iscoroutine(coro) or isinstance(coro, GeneratorType):
-            return _cancellableInlineCallbacks(coro)
-
-    if not isinstance(coro, Deferred):
-        raise ValueError("%r is not a coroutine or a Deferred" % (coro,))
-
-    # Must be a Deferred
-    return coro
+    if isinstance(coro, Deferred):
+        return coro
+    else:
+        try:
+            return Deferred.fromCoroutine(coro)
+        except ValueError:
+            # It's not a coroutine. Raise an exception, but say that it's also
+            # not a Deferred so the error makes sense.
+            raise ValueError("%r is not a coroutine or a Deferred" % (coro,))
```

Comment:
I think you want raise ... from (the other exception just raised) for clarity when debugging? Just a suggestion though, feel free to override. Also should be a new exception type though, derived from TypeError not ValueError

Reply:
NotACoroutineError added, subclass of TypeError

Reply:
@glyph: Not sure if you meant more than [fixing the type](https://github.com/twisted/twisted/pull/1269/files#diff-0a695d3d59d1e9452e21eb981b3bb687R958)